### PR TITLE
Added an optional frameswritten setter

### PR DESCRIPTION
**Pull request by evalott100**
Added an optional setter in the HDF5 writer to set a record detailing the number of frames written.

We require merging and release of this PR before the ioc side: https://github.com/PandABlocks/PandABlocks-ioc/pull/83

According to https://github.com/dls-controls/pythonSoftIOC/issues/142, the set of a record should be thread safe, but we should do some testing before we release.

### CODE REVIEW

File: src/pandablocks/hdf.py
Comment (coretl) [inline src/pandablocks/hdf.py:1]: An alternative idea for you, at the moment the result of `self.write_frame(data)` is pushed to `self.downstream.queue`. If we return `dataset.shape[0]` from `self.write_frame`, then in PandABlocks-ioc we could create a new `Pipeline` subclass that took the frames written integer and called `record.set()` with the value

Comment (coretl) [reply src/pandablocks/hdf.py:1]: This would be minimal changes here, and be completely backwards compatible...

Comment (evalott100) [reply src/pandablocks/hdf.py:1]: That's nicer, I'll implement.

File: src/pandablocks/hdf.py
Comment (coretl) [inline src/pandablocks/hdf.py:86]: Can we remove this trailing comma please?

File: src/pandablocks/hdf.py
Comment (coretl) [inline src/pandablocks/hdf.py:172]: Why do you need this?

**Review by coretl — COMMENTED** [review]
Couple of minor comments, can we write a test that the file number comes out of the end of the pipeline?

File: src/pandablocks/hdf.py
Comment (evalott100) [reply src/pandablocks/hdf.py]: Accidentally left this in!

**Comment by evalott100** [comment]
@coretl completely forgot about adding the test till now. dmgav approved the ioc side add_ringbuffer PR so after this is released I'll merge and release that if CI is all good?

File: src/pandablocks/responses.py
Comment (coretl) [inline src/pandablocks/responses.py:200]: Can you put these at the bottom, after `START_DATA_MISMATCH` as they are not generated by the server

**Review by coretl — APPROVED** [review]

**Pull request merged**
